src/transforms/geoip.rs

```diff
@@ -0,0 +1,256 @@
+use super::Transform;
+
+use crate::{
+    event::Event,
+    topology::config::{DataType, TransformConfig},
+};
+use serde::{Deserialize, Serialize};
+use string_cache::DefaultAtom as Atom;
+
+use indexmap::IndexMap;
+use std::net::IpAddr;
+use std::str::FromStr;
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct GeoipConfig {
+    pub source: Atom,
+    pub database: String,
+    #[serde(default = "default_geoip_target_field")]
+    pub target: String,
+}
+
+pub struct Geoip {
+    pub dbreader: maxminddb::Reader<Vec<u8>>,
+    pub source: Atom,
+    pub target: String,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct GeoipDecodedData {
+    pub city_name: String,
+    pub continent_code: String,
+    pub country_code: String,
+    pub time_zone: String,
+    pub latitude: String,
+    pub longitude: String,
+    pub postal_code: String,
+}
+
+fn default_geoip_target_field() -> String {
+    "geoip".to_string()
+}
+
+#[typetag::serde(name = "geoip")]
+impl TransformConfig for GeoipConfig {
+    fn build(&self) -> Result<Box<dyn Transform>, crate::Error> {
+        let reader = maxminddb::Reader::open_readfile(self.database.clone()).unwrap();
+        Ok(Box::new(Geoip::new(
+            reader,
+            self.source.clone(),
+            self.target.clone(),
+        )))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+}
+
+impl Geoip {
+    pub fn new(dbreader: maxminddb::Reader<Vec<u8>>, source: Atom, target: String) -> Self {
+        Geoip {
+            dbreader: dbreader,
+            source: source,
+            target: target,
+        }
+    }
+}
+
+impl Transform for Geoip {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        let ipaddress = event
+            .as_log()
+            .get(&self.source)
+            .map(|s| s.to_string_lossy());
+        if let Some(ipaddress) = &ipaddress {
+            let mut lookup_results = IndexMap::new();
+            let ip: IpAddr = FromStr::from_str(ipaddress).unwrap();
+            let v = self.dbreader.lookup(ip);
+            if v.is_ok() {
+                let data: maxminddb::geoip2::City = v.unwrap();
```

Comment:
This should probably be collapsed down to something like this: 
```rust
if let (data) = self.dbreader.lookup(ip) {
```

Reply:
Can you please explain this a bit more? I am very much a Rust beginner, hence thanks in advance for your explanations.

Reply:
Sorry, I think I had a typo in there. The following are equivalent:

```rust
let v = self.dbreader.lookup(ip);
if v.is_ok() {
    let data: maxminddb::geoip2::City = v.unwrap();
    ...
```

and

```rust
if let Ok(data) = self.dbreader.lookup(ip) {
    ...
```

The `if let` construct is a more concise way of saying "if this value matches the pattern, destructure and operate on it".

You may need to add a type annotation to `lookup` in the second case (`self.dbreader.lookup::<maxminddb::geoip2::City>(ip)`) to replace the annotation in the first on `data`.